example/src/main/java/io/netty/example/stomp/websocket/StompChatHandler.java

```diff
@@ -0,0 +1,193 @@
+import static io.netty.handler.codec.stomp.StompHeaders.*;
+import static io.netty.util.CharsetUtil.*;
+
+@Sharable
+public class StompChatHandler extends SimpleChannelInboundHandler<StompFrame> {
+
+    static final StompChatHandler INSTANCE = new StompChatHandler();
+
+    private final ConcurrentHashMap<String, Set<StompSubscription>> CHAT_DESTINATIONS =
+            new ConcurrentHashMap<String, Set<StompSubscription>>();
+
+    @Override
+    protected void channelRead0(ChannelHandlerContext ctx, StompFrame inboundFrame) throws Exception {
+        DecoderResult decoderResult = inboundFrame.decoderResult();
+        if (decoderResult.isFailure()) {
+            sendErrorFrame("rejected frame", decoderResult.toString(), ctx);
+            return;
+        }
+
+        switch (inboundFrame.command()) {
+        case STOMP:
+        case CONNECT:
+            onConnect(inboundFrame, ctx);
+            break;
+        case SUBSCRIBE:
+            onSubscribe(inboundFrame, ctx);
+            break;
+        case SEND:
+            onSend(inboundFrame, ctx);
+            break;
+        case UNSUBSCRIBE:
+            onUnsubscribe(inboundFrame, ctx);
+            break;
+        case DISCONNECT:
+            onDisconnect(inboundFrame, ctx);
+            break;
+        default:
+            sendErrorFrame("unsupported command",
+                           "Received unsupported command " + inboundFrame.command(), ctx);
+        }
+    }
+
+    private void onConnect(StompFrame inboundFrame, ChannelHandlerContext ctx) {
+        String acceptVersions = inboundFrame.headers().getAsString(ACCEPT_VERSION);
+        StompVersion handshakeAcceptVersion = ctx.channel().attr(StompVersion.CHANNEL_ATTRIBUTE_KEY).get();
+        if (acceptVersions == null || !acceptVersions.contains(handshakeAcceptVersion.version())) {
+            sendErrorFrame("invalid version",
+                           "Received invalid version, expected " + handshakeAcceptVersion.version(), ctx);
+            return;
+        }
+
+        StompFrame connectedFrame = new DefaultStompFrame(StompCommand.CONNECTED);
+        connectedFrame.headers()
+                      .set(VERSION, handshakeAcceptVersion.version())
+                      .set(SERVER, "Netty-Server")
+                      .set(HEART_BEAT, "0,0");
+        ctx.writeAndFlush(connectedFrame);
+    }
+
+    private void onSubscribe(StompFrame inboundFrame, ChannelHandlerContext ctx) {
+        String destination = inboundFrame.headers().getAsString(DESTINATION);
+        String subscriptionId = inboundFrame.headers().getAsString(ID);
+
+        if (destination == null || subscriptionId == null) {
+            sendErrorFrame("missed header", "Required 'destination' or 'id' header missed", ctx);
+            return;
+        }
+
+        Set<StompSubscription> subscriptions =
+                CHAT_DESTINATIONS.putIfAbsent(destination, new HashSet<StompSubscription>());
+        if (subscriptions == null) {
```

Comment:
can this ever be null ? I mean you use ´.putIfAbsent´ just before.

Reply:
If this first attempt it's return null (`map.put(k,v)`)